rmf_gazebo_plugins/src/readonly.cpp

```diff
@@ -0,0 +1,466 @@
+    RCLCPP_INFO(logger(), "Start waypoint successfully initialized");
+  }
+
+  else
+  {
+    RCLCPP_ERROR(
+      logger(),
+      "Spawn coordinates [%f,%f,%f] differs from that of waypoint [%s] in nav_graph [%f, %f, %f]",
+       pose.Pos().X(), pose.Pos().Y(),0,
+      _start_wp_name.c_str(),
+      _graph.vertices[_start_wp].x, _graph.vertices[_start_wp].y, 0);
+  }
+  
+}
+std::size_t ReadonlyPlugin::get_next_waypoint(const std::size_t& start_wp,
+    const ignition::math::Vector3d& heading,
+    const ignition::math::Pose3d& pose)
+{
+  // Return the waypoint closest to the robot in the direction of its heading
+  const auto& neighbors = _neighbor_map.find(start_wp)->second;
+
+  auto wp_it = neighbors.begin();
+  double max_dist = std::numeric_limits<double>::min();
+
+  for (auto it = neighbors.begin(); it != neighbors.end(); it++)
+  {
+    const auto& waypoint = _graph.vertices[*it];
+    ignition::math::Vector3d disp_vector{
+        waypoint.x - _graph.vertices[start_wp].x,
+        waypoint.y - _graph.vertices[start_wp].y,
+        0};
+    const double dist = heading.Dot(disp_vector.Normalize());
+    // Consider the waypoints with highest projected distance
+    if (dist > max_dist)
+    {
+      max_dist = dist;
+      wp_it = it;
+    }
+  }
+
+  return *wp_it;
+}
+
+ReadonlyPlugin::Path ReadonlyPlugin::compute_path(const ignition::math::Pose3d& pose)
+{
+  Path path;
+  path.resize(_lookahead);
+
+  std::size_t count = 0;
+  
+  auto start_wp = _start_wp;
+  const double current_yaw = pose.Rot().Euler().Z();
+  ignition::math::Vector3d heading{
+      std::cos(current_yaw), std::sin(current_yaw), 0.0};
+  
+  while (count < _lookahead)
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    auto wp = get_next_waypoint(start_wp, heading, pose);
+    _next_wp[count] = wp;
+    // Add to path here
+    Location location;
+    location.x = _graph.vertices[wp].x;
+    location.y = _graph.vertices[wp].y;
+    location.level_name = _level_name;
+    path[count] = location;
+    // Update heading for next iteration
+    auto next_heading = ignition::math::Vector3d{
+      _graph.vertices[wp].x - _graph.vertices[start_wp].x,
+      _graph.vertices[wp].y - _graph.vertices[start_wp].y,
+      0};
+    heading = next_heading.Normalize();
+    start_wp = wp;
+    ++count;
+  }
+
+  return path;
+
+}
+
+void ReadonlyPlugin::OnUpdate()
+{
+  _update_count++;
+  const auto& world = _model->GetWorld();
+  auto pose = _model->WorldPose();
+
+  if (_update_count % 100 == 0) // todo: be smarter, use elapsed sim time
+  {
+    initialize_start(pose);
+
+    const double time = world->SimTime().Double();
+    _last_update_time = time;
+    const int32_t t_sec = static_cast<int32_t>(time);
+    const uint32_t t_nsec =
+        static_cast<uint32_t>((time-static_cast<double>(t_sec)) *1e9);
+    const rclcpp::Time now{t_sec, t_nsec, RCL_ROS_TIME};
+
+    _robot_state_msg.name = _model->GetName();
+    _robot_state_msg.model = "";
```

Comment:
we can probably set/get the model type from the sdf elements as well, this will let us use this plugin for all sorts of mobile platforms in the future :boom:

Reply:
Will keep in this in mind for the next iteration.